Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnCancel.hs

```diff
@@ -0,0 +1,128 @@
+
+ is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+
+ or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details. You should have received a copy of
+
+ the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+module Beckn.ACL.OnCancel where
+
+import qualified Beckn.Types.Core.Taxi.Common.Tags as Tags
+import qualified Beckn.Types.Core.Taxi.OnCancel as OnCancel
+import Data.Text
+import qualified Domain.Types.Booking as DRB
+import qualified Domain.Types.BookingCancellationReason as SBCR
+import qualified Domain.Types.Ride as R
+import qualified Domain.Types.SearchRequest as DSR
+import EulerHS.Prelude hiding (pack)
+import Kernel.External.Encryption
+import Kernel.Storage.Esqueleto
+import Kernel.Types.Error
+import Kernel.Types.Id
+import Kernel.Utils.Error
+import Storage.Queries.Estimate as QE
+import Storage.Queries.SearchRequest as QS
+import Tools.Error
+
+data OnCancelBuildReq = OnCancelBuildReq
+  { booking :: DRB.Booking,
+    mbRideId :: Maybe (Id R.Ride),
+    cancellationSource :: SBCR.CancellationSource
+  }
+
+newtype OnSearchCancelBuildReq = OnSearchCancelBuildReq {searchReqId :: Id DSR.SearchRequest}
+
+buildOnCancelMessage :: (EsqDBFlow m r, EncFlow m r) => OnCancelBuildReq -> m OnCancel.OnCancelMessage
+buildOnCancelMessage OnCancelBuildReq {..} = do
+  fId <- getFulFillmentId booking mbRideId
+  let tagGroups =
+        Tags.TG
+          [ Tags.TagGroup
+              { display = True,
+                code = "booking_cancelled_info",
+                name = "Booking Cancelled Info",
+                list = [Tags.Tag (Just True) (Just "booking_cancelled") (Just "Booking Cancelled") (Just $ show True)]
+              },
+            Tags.TagGroup
+              { display = True,
+                code = "cancellation_source_info",
+                name = "Cancellation Source Info",
+                list = [Tags.Tag (Just True) (Just "cancellation_source") (Just $ show cancellationSource) (Just $ show cancellationSource)]
+              }
+          ]
+  return $
+    OnCancel.OnCancelMessage $
+      OnCancel.Order
+        { id = Just booking.id.getId,
+          state = "CANCELLED",
+          fulfillment =
+            OnCancel.FulfillmentInfo
+              { id = fId,
+                state =
+                  Just
+                    OnCancel.FulfillmentState
+                      { descriptor =
+                          OnCancel.Descriptor
+                            { short_desc = Nothing,
+                              code = Just "CANCELLED"
+                            }
+                      },
+                tags = Just tagGroups
+              }
+        }
+
+getFulFillmentId :: (EsqDBFlow m r, EncFlow m r) => DRB.Booking -> Maybe (Id R.Ride) -> m Text
```

Comment:
Very confusing logic for define `fulfillmentId`. How we should handle it on bap side? Can we just make it optional and equal to `Just rideId `(or `Nothing` if no ride exist)? the same we have in `on_update`